### PR TITLE
Add some unit tests and integration tests, fix a new-pat arg issue

### PR DESCRIPTION
**Pull request by Exolun**
This pull request includes a bunch of end to end tests and unit tests.

The end to end tests work by prompting the user to enter a series of Unity directories that exercise the test scenarios, and an azure subscription GUID to test the new AzureSubscription arg.  It is by no means comprehensive, but it should more or less covers all of the valid happy paths.  At the time of this PR there are 15 of these tests.

Similarly there is also a body of unit tests - these tests are mostly identical to the e2e tests except they use mocks heavily.  Realistically the unit tests really don't test much because of how little of the underlying code actually runs.  In fact, due to the support of `$PSCmdlet.ShouldProcess`, I found it nigh-impossible to even mock/test `Update-PackageAuthConfig`, which is where most of the interesting code paths lie.

Pester was fairly easy to work with, but I did find that the execution scope that the tests are run within makes it quite challenging to reuse code and do anything elaborate with the mocking - hence the lack of shared code across the two test files.

### CODE REVIEW

File: Tests/UnitTests/Update-UnityPackageManagerConfig.Tests.ps1 [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+        }
+
+        It 'throws on null ProjectManifestPath' {
+            { Update-UnityPackageManagerConfig -ProjectManifestPath "" } | Should -Throw "*The argument is null or empty.*"
+        }
+    }
+
+    Context 'Function Execution' {
+        It 'supports a root folder target' {
+            { Update-UnityPackageManagerConfig -SearchPath ".\Data\" -SearchDepth 5 } | Should -Not -Throw
```
Comment (jwittner) [inline Tests/UnitTests/Update-UnityPackageManagerConfig.Tests.ps1:31]: Can we modify these relative paths to use $PSScriptRoot? Right now if I don't cd into the 'Tests' folder then I get errors running the unittests.ps1 script.

File: Tests/unittests.ps1 [hunk 2]
```diff
@@ -0,0 +1,24 @@
+#Requires -Modules @{ModuleName = 'PSSCriptAnalyzer'; ModuleVersion = '1.20.0'}
```
Comment (jwittner) [inline Tests/unittests.ps1:1]: Can we add a README to this directory about how to use these new scripts and write new tests? And link to it from the root README?

**Review by jwittner — CHANGES_REQUESTED** [review]
A couple of small requests, but this is an AWESOME contribution to lay the foundation for improved testing!

File: Tests/UnitTests/Update-UnityPackageManagerConfig.Tests.ps1 (hunk 1 above)
Comment (Exolun) [reply Tests/UnitTests/Update-UnityPackageManagerConfig.Tests.ps1:31]: Good call - will update this

File: Tests/unittests.ps1 (hunk 2 above)
Comment (Exolun) [reply Tests/unittests.ps1:1]: Great idea - I'll update the docs!

**Review by jwittner — APPROVED** [review]
Updates look awesome! Only nit would be to ignore the TestResults.xml from any path so that we can run the script from anywhere.

**Comment by Exolun** [comment]
> Updates look awesome! Only nit would be to ignore the TestResults.xml from any path so that we can run the script from anywhere.

Sounds good - since I will be OOF next week I think I'll merge this and create a branch withTestResults.xml change to gitignore for a fast-follow up.

**Pull request merged**